README.md
<p align="center">
    <img src="res/synth.png" alt="Castle Quilldore">
</p>
<h3 align="center">Web Audio API Synth</h3>
<hr>
<p align="center">
  <a href="../../issues">Report an Issue</a>
  ·
  <a href="../../releases">Releases</a>
  ·
  <a href="/CHANGELOG.md">Change Log</a>
</p>
<p align="center">
  <a href="#Description">Description</a>
  ·
  <a href="#Roadmap">Roadmap</a>
</p>
<hr>

## Description
Synth made from Web Audio API. Using [Tuna.js](https://github.com/Theodeus/tuna), [Kontrol.js](http://anthonyterrien.com/kontrol/) and [Qwerty.js](https://stuartmemo.com/qwerty-hancock/)

## Roadmap
Date | Issue | Status |
:--: | ----- | ------ |
20/05/2021 | Add start button to stop "The AudioContext was not allowed to start. It must be resumed (or created) after a user gesture on the page" warning | complete - 22/05/2021 |
20/05/2021 | Drum machine | completed - 22/05/2021 |
20/05/2021 | Make mobile friendly | completed - 25/05/2021 |
20/05/2021 | Refactor, refactor, refactor | completed - 22/05/2021 |
25/05/2021 | Midi effect control | incomplete |


CHANGELOG.md
# Change Log
## Web Audio API Synth
Version | Date | Developer | Changes |
------- | ---- | --------- | ------- |
1.0.0 | 20/05/2021 | Toby Butchart | Initial release |
1.1.0 | 22/05/2021 | Toby Butchart | <ul><li>Added modal popup on start</li><li>Added drum machine</li><li>Refactored js code into modules</li></ul> |
1.2.0 | 23/05/2021 | Toby Butchart | <ul><li>Removed keydown and keyup events from Qwerty.js so the drum machine is played by the keyboard (was playing both keyboard and drums)</li><li>Added click and touch events to drum pads so it's playable on a phone</li></ul> |
1.2.1 | 23/05/2021 | Toby Butchart | Turns out a touch event isn't needed... |
1.3.0 | 25/05/2021 | Toby Butchart | Made mobile firendly |
1.3.1 | 25/05/2021 | Toby Butchart | <ul><li>Shrunk control font-size if small screen</li><li>Changes keyboard octave count depending on screen size (1, 2 or 3 octaves)</li></ul> |


src/js/drmz.js
/*maps keyboard letters to drum sounds*/
var keyboardMap = {
    'q':36,
    'w':37,
    'e':38,
    'a':39,
    's':40,
    'd':41,
    'z':42,
    'x':43,
    'c':44
};

function createDrumMachine(){
    document.getElementById("drum-machine").style.display = "block";
}

function playDrumSound(note){
    const drmAudioElem = document.querySelector('audio[data-key="'+note+'"]');
    playDrumSoundFromElement(drmAudioElem);
}

function playDrumSoundFromElement(e){
    //if from KeyboardEvent
    if(e && e.key && keyboardMap[e.key]){
        e = document.querySelector('audio[data-key="'+keyboardMap[e.key]+'"]');
    }

    //if from MouseEvent
    if(e && e.srcElement && e.srcElement.dataset && e.srcElement.dataset.key){
        e = document.querySelector('audio[data-key="'+e.srcElement.dataset.key+'"]');
    }

    if(!e || !e.dataset || !e.dataset.key){
        return;
    }

    const key = document.querySelector('.key[data-key="'+e.dataset.key+'"]');

    if(!key || !key.dataset || !key.dataset.key){
        return;
    }

    if(!drmAudio["key-"+key.dataset.key]){
        drmAudio["key-"+key.dataset.key] = audioContext.createMediaElementSource(e);
    }

    createPatching();

    e.currentTime = 0;
    e.play();

    key.classList.add('playing');
}

function removeTransition(e){
    if(e.propertyName !== 'transform'){
        return;
    }
    this.classList.remove('playing');
}

window.addEventListener('keydown', playDrumSoundFromElement);

const keys = document.querySelectorAll('.key');
for(var i = 0; i < keys.length; i++){
    keys[i].addEventListener('transitionend', removeTransition);
    keys[i].addEventListener('click', playDrumSoundFromElement);
    // keys[i].addEventListener('touchstart', playDrumSoundFromElement);
}


src/js/midi.js
/*https://gist.github.com/YuxiUx/ef84328d95b10d0fcbf537de77b936cd*/
/*I had to rewrite this as my stupid minifyer didn't like the exponentiation ** operator*/
function midiNoteToFreq(note) {
    var a = 440;
    var b = (a / 32);
    var c = ((note - 9) / 12);
    return b * Math.pow(2, c);
}

function freqToMidiNote(frequency) {
     return Math.round(69 + 12 * Math.log2(frequency / 440));
}

function setMIDI() {
    navigator.requestMIDIAccess().then(onMIDISuccess, onMIDIFailure);
}

function onMIDIFailure() {
    console.log('Could not access your MIDI devices :(');
}

function onMIDISuccess(midiAccess) {
    midiAccess.inputs.forEach(
        function(item){
            item.onmidimessage = getMIDIMessage;
        }
    );
}

function getMIDIMessage(message) {
    var command = message.data[0];
    var note = message.data[1];
    var velocity = (message.data.length > 2) ? message.data[2] : 0; // a velocity value might not be included with a noteOff command

    switch (command) {
        case 144: // noteOn
            if (velocity > 0) {
                MIDINoteOn(note, velocity);
            } else {
                MIDINoteOff(note);
            }
            break;
        case 128: // noteOff
            MIDINoteOff(note);
            break;
    }
}

function MIDINoteOn(note) {
    var frequency = midiNoteToFreq(note);
    if(note < 36 || note > 44){
        startNote('', frequency);
    }else{
        playDrumSound(note);
    }
}

function MIDINoteOff(note) {
    var frequency = midiNoteToFreq(note);
    stopNote('', frequency);
}


src/js/synth.js
function createSynth(){
    const keyboardNode = document.getElementById("keyboard");
    keyboardNode.innerHTML = '';

    const parentWidth = $('#keyboard-enclosure').width();
    var octaveCount;

    if(parentWidth < 576){
        octaveCount = 1;
    }else if(parentWidth < 992){
        octaveCount = 2;
    }else{
        octaveCount = 3;
    }

    keyboard = new QwertyHancock({
         id: 'keyboard',
         width: parentWidth,
         height: 250,
         octaves: octaveCount,
         startNote: 'C4'
    });

    document.getElementById("synth-effects").style.display = "block";
}

function animateBar(barElement, withTimeout, timeout, oscIndex){
    var v =  0;
    var s = 'up';
    var i = 0;

    var redraw = function () {
        if(v == 15 || v == 0){
            i++;
        }

        s = v == 15 ? 'down' : (v == 0 ? 'up' : s);
        s == 'up' ? v++ : v--;

        barElement
            .find('input:eq('+v+')')
            .val(Math.floor(Math.sin(i+v*i)*1000))
            .trigger('change');

        if(canLoop[oscIndex]){
            timeout = window.setTimeout(redraw, 1000/60);
        }
    };

    redraw();

    if(withTimeout){
        window.setTimeout(function(){ window.clearTimeout(timeout); }, 1000);
    }
}


src/js/events.js
$(window).resize(function(){
    var parentWidth = $('#div-custom-osc1').width();
    $('#bars-custom-osc1').trigger("configure", {"width":parentWidth});

    var parentWidth = $('#div-custom-osc2').width();
    $('#bars-custom-osc2').trigger("configure", {"width":parentWidth});

    var parentWidth = $('#div-lfo1').width();
    $('#xy-lfo1').trigger("configure", {"width":parentWidth});

    resizeEffectDials();
    resizeXYBar();
    createSynth();
});

function resizeEffectDials(){
    if($( window ).width() > 567){
        $('.effect-dial').trigger('configure',{"width":80, "height":80});
    }else{
        $('.effect-dial').trigger('configure',{"width":60, "height":60});
    }
}

function resizeXYBar(){
    var parentWidth = $('#div-lfo1').width();
    $('#xy-lfo1').trigger("configure", {width:parentWidth});
}

function setOsc(waveform, osc, barsID, frequency){
    osc.frequency.setTargetAtTime(frequency, audioContext.currentTime, 0);

    if(waveform === 'custom'){
        var f = [];

        $(barsID).each(function(index, element) {
            f.push(parseFloat($(element).val() / 1000));
        });

        sineTerms = new Float32Array(f);
        cosineTerms = new Float32Array(sineTerms.length);
        customWaveform = audioContext.createPeriodicWave(cosineTerms, sineTerms);

        osc.setPeriodicWave(customWaveform);
    }else{
        osc.type = waveform;
    }
}

function startNote(note, frequency){
    osc1 = audioContext.createOscillator();
    osc2 = audioContext.createOscillator();
    lfo1 = audioContext.createOscillator();

    setOsc(document.getElementById("osc1Type").value, osc1, '#bars-custom-osc1 :input', frequency);
    setOsc(document.getElementById("osc2Type").value, osc2, '#bars-custom-osc2 :input', frequency);

    lfo1.frequency.setTargetAtTime(parseFloat($('#lfo1-freq').val() / 5), audioContext.currentTime, 0);
    lfo1.type = document.getElementById("lfo1Type").value;

    createPatching();

    oscillators[frequency] = [osc1, osc2, lfo1];

    osc1.detune.value = document.getElementById("dial-finetune-osc1").value;
    osc2.detune.value = document.getElementById("dial-finetune-osc2").value;

    osc1.start(audioContext.currentTime);
    osc2.start(audioContext.currentTime);
    lfo1.start(audioContext.currentTime);
}

function stopNote(note, frequency){
    if(oscillators[frequency]){
        oscillators[frequency].forEach(function (oscillator) {
            oscillator.stop(audioContext.currentTime);
        });
    }
}

function setEvents(){
    $(".effect-dial").knob({
        'release' : function (v) {
            var e = this.$[0];
            var effectIndex = parseInt(this.$[0].dataset.effectindex);
            var controlIndex = parseInt(this.$[0].dataset.controlindex);
            var effect = window[effects[effectIndex].effectType];

            effect[effects[effectIndex].controls[controlIndex].controlProperty] = v;
        }
    });

    resizeEffectDials();
    resizeXYBar();

    keyboard.keyDown = function (note, frequency) {
        startNote(note, frequency);
        playDrumSound(note);
    };

    keyboard.keyUp = function (note, frequency) {
        stopNote(note, frequency);
    };

    $(".dial").dial();

    $(".bars").bars({
        'width':545,
        'cols':16,
        'min':-100,
        'max':100,
        'fgColor':"#FFF"
    });

    $(".xy").xy();

    var timeoutOSC1;
    var timeoutOSC2;

    $("#randomise-osc1").bind('click',
        function() {
            animateOSC1 = true;
            animateBar($(".bar-osc1"), true, timeoutOSC1, 1);
        }
    );

    $("#start-osc1").bind('click',
        function() {
            canLoop[1] = true;
            animateBar($(".bar-osc1"), false, timeoutOSC1, 1);
        }
    );

    $("#stop-osc1").bind('click',
        function() {
            canLoop[1] = false;
        }
    );

    $("#randomise-osc2").bind('click',
        function() {
            animateOSC2 = true;
            animateBar($(".bar-osc2"), true, timeoutOSC2, 2);
        }
    );

    $("#start-osc2").bind('click',
        function() {
            canLoop[2] = true;
            animateBar($(".bar-osc2"), false, timeoutOSC2, 2);
        }
    );

    $("#stop-osc2").bind('click',
        function() {
            canLoop[2] = false;
        }
    );
}


src/js/qwerty-alt.js
/* Qwerty Hancock 0.6.0 (c) 2012-18 Stuart Memo */
!function() {
    var a = this
      , b = "undefined" == typeof global ? a : a.window
      , c = "0.6.0"
      , d = {}
      , e = !1
      , f = {}
      , g = {
        65: "Cl",
        87: "C#l",
        83: "Dl",
        69: "D#l",
        68: "El",
        70: "Fl",
        84: "F#l",
        71: "Gl",
        89: "G#l",
        90: "G#l",
        72: "Al",
        85: "A#l",
        74: "Bl",
        75: "Cu",
        79: "C#u",
        76: "Du",
        80: "D#u",
        59: "Eu",
        186: "Eu",
        222: "Fu",
        221: "F#u",
        220: "Gu"
    }
      , h = function(a) {
        return Math.floor((d.width - a) / a)
    }
      , i = function(a) {
        var b;
        user_settings = a || {},
        d = {
            id: user_settings.id || "keyboard",
            octaves: user_settings.octaves || 3,
            width: user_settings.width,
            height: user_settings.height,
            startNote: user_settings.startNote || "A3",
            whiteKeyColour: user_settings.whiteKeyColour || "#fff",
            blackKeyColour: user_settings.blackKeyColour || "#000",
            activeColour: user_settings.activeColour || "yellow",
            borderColour: user_settings.borderColour || "#000",
            keyboardLayout: user_settings.keyboardLayout || "en"
        },
        b = document.getElementById(d.id),
        "undefined" == typeof d.width && (d.width = b.offsetWidth),
        "undefined" == typeof d.height && (d.height = b.offsetHeight),
        d.startOctave = parseInt(d.startNote.charAt(1), 10),
        A(),
        F.call(this, b)
    }
      , j = function(a) {
        var b, c, d = ["A", "A#", "B", "C", "C#", "D", "D#", "E", "F", "F#", "G", "G#"];
        return c = 3 === a.length ? a.charAt(2) : a.charAt(1),
        b = d.indexOf(a.slice(0, -1)),
        b = 3 > b ? b + 12 + 12 * (c - 1) + 1 : b + 12 * (c - 1) + 1,
        440 * Math.pow(2, (b - 49) / 12)
    }
      , k = function(a) {
        (null !== a || void 0 === typeof a) && (a.style.backgroundColor = d.activeColour)
    }
      , l = function(a) {
        null !== a && ("white" === a.getAttribute("data-note-type") ? a.style.backgroundColor = d.whiteKeyColour : a.style.backgroundColor = d.blackKeyColour)
    }
      , m = function(a) {
        var b, c = 0, e = a.length, f = [];
        for (b = 0; e > b; b++)
            if (d.startNote.charAt(0) === a[b]) {
                c = b;
                break
            }
        for (b = 0; e > b; b++)
            b + c > e - 1 ? f[b] = a[b + c - e] : f[b] = a[b + c];
        return f
    }
      , n = function(a) {
        a.el.style.backgroundColor = d.whiteKeyColour,
        a.el.style.border = "1px solid " + d.borderColour,
        a.el.style.borderRight = 0,
        a.el.style.height = d.height + "px",
        a.el.style.width = a.width + "px",
        a.el.style.borderRadius = "0 0 5px 5px",
        a.noteNumber === w() - 1 && (a.el.style.border = "1px solid " + d.borderColour)
    }
      , o = function(a) {
        var b = h(w())
          , c = Math.floor(b / 2);
        a.el.style.backgroundColor = d.blackKeyColour,
        a.el.style.border = "1px solid " + d.borderColour,
        a.el.style.position = "absolute",
        a.el.style.left = Math.floor((b + 1) * (a.noteNumber + 1) - c / 2) + "px",
        a.el.style.width = c + "px",
        a.el.style.height = d.height / 1.5 + "px",
        a.el.style.borderRadius = "0 0 3px 3px"
    }
      , p = function(a) {
        a.el.style.display = "inline-block",
        a.el.style["-webkit-user-select"] = "none",
        "white" === a.colour ? n(a) : o(a)
    }
      , q = function(a) {
        var b = function(a) {
            a.style.cursor = "default",
            a.style.fontSize = "0px",
            a.style.height = d.height + "px",
            a.style.padding = 0,
            a.style.position = "relative",
            a.style.listStyle = "none",
            a.style.margin = 0,
            a.style.width = d.width + "px",
            a.style["-webkit-user-select"] = "none"
        };
        b(a.container),
        b(a.el)
    }
      , r = function(a, b) {
        "li" == a.tagName.toLowerCase() && (e = !0,
        k(a),
        b(a.title, j(a.title)))
    }
      , s = function(a, b) {
        "li" == a.tagName.toLowerCase() && (e = !1,
        l(a),
        b(a.title, j(a.title)))
    }
      , t = function(a, b) {
        e && (k(a),
        b(a.title, j(a.title)))
    }
      , u = function(a, b) {
        e && (l(a),
        b(a.title, j(a.title)))
    }
      , v = function(a) {
        return a.el = document.createElement("li"),
        a.el.id = a.id,
        a.el.title = a.id,
        a.el.setAttribute("data-note-type", a.colour),
        p(a),
        a
    }
      , w = function() {
        return 7 * d.octaves
    }
      , x = function() {
        var a, b, c = this, e = [], f = 0, g = d.startOctave, i = w();
        for (a = 0; i > a; a++)
            a % this.whiteNotes.length === 0 && (f = 0),
            bizarre_note_counter = this.whiteNotes[f],
            "C" === bizarre_note_counter && 0 !== a && g++,
            b = v({
                colour: "white",
                octave: g,
                width: h(i),
                id: this.whiteNotes[f] + g,
                noteNumber: a
            }),
            e.push(b.el),
            a !== i - 1 && this.notesWithSharps.forEach(function(d, j) {
                d === c.whiteNotes[f] && (b = v({
                    colour: "black",
                    octave: g,
                    width: h(i) / 2,
                    id: c.whiteNotes[f] + "#" + g,
                    noteNumber: a
                }),
                e.push(b.el))
            }),
            f++;
        return e
    }
      , y = function(a) {
        a.keys.forEach(function(b) {
            a.el.appendChild(b)
        })
    }
      , z = function(a) {
        d.keyPressOffset = "C" === a[0] ? 0 : 1
    }
      , A = function() {
        var a = {
            container: document.getElementById(d.id),
            el: document.createElement("ul"),
            whiteNotes: m(["C", "D", "E", "F", "G", "A", "B"]),
            notesWithSharps: m(["C", "D", "F", "G", "A"])
        };
        return a.keys = x.call(a),
        z(a.whiteNotes),
        q(a),
        y(a),
        a.container.appendChild(a.el),
        a
    }
      , B = function(a) {
        return g[a].replace("l", parseInt(d.startOctave, 10) + d.keyPressOffset).replace("u", (parseInt(d.startOctave, 10) + d.keyPressOffset + 1).toString())
    }
      , C = function(a, b) {
        var c;
        return a.keyCode in f ? !1 : (f[a.keyCode] = !0,
        "undefined" != typeof g[a.keyCode] ? (c = B(a.keyCode),
        b(c, j(c)),
        k(document.getElementById(c)),
        !0) : !1)
    }
      , D = function(a, b) {
        var c;
        return delete f[a.keyCode],
        "undefined" != typeof g[a.keyCode] ? (c = B(a.keyCode),
        b(c, j(c)),
        l(document.getElementById(c)),
        !0) : !1
    }
      , E = function(a) {
        return a.ctrlKey || a.metaKey || a.altKey
    }
      , F = function(a) {
        var c = this;
        // b.addEventListener("keydown", function(a) {
        //     E(a) || C(a, c.keyDown) && a.preventDefault()
        // }),
        // b.addEventListener("keyup", function(a) {
        //     E(a) || D(a, c.keyUp) && a.preventDefault()
        // }),
        a.addEventListener("mousedown", function(a) {
            r(a.target, c.keyDown)
        }),
        a.addEventListener("mouseup", function(a) {
            s(a.target, c.keyUp)
        }),
        a.addEventListener("mouseover", function(a) {
            t(a.target, c.keyDown)
        }),
        a.addEventListener("mouseout", function(a) {
            u(a.target, c.keyUp)
        }),
        "ontouchstart"in document.documentElement && (a.addEventListener("touchstart", function(a) {
            r(a.target, c.keyDown)
        }),
        a.addEventListener("touchend", function(a) {
            s(a.target, c.keyUp)
        }),
        a.addEventListener("touchleave", function(a) {
            u(a.target, c.keyUp)
        }),
        a.addEventListener("touchcancel", function(a) {
            u(a.target, c.keyUp)
        }))
    }
      , G = function(a) {
        this.version = c,
        this.keyDown = function() {}
        ,
        this.keyUp = function() {}
        ,
        i.call(this, a)
    };
    "undefined" != typeof exports ? ("undefined" != typeof module && module.exports && (exports = module.exports = G),
    exports.QwertyHancock = G) : a.QwertyHancock = G
}(this);
